main/src/com/google/refine/expr/functions/strings/ParseUri.java

```diff
@@ -43,30 +39,28 @@ public Object call(Properties bindings, Object[] args) {
                     }
                 }
 
-                // initial
-                Map<String, String> queryParams = new HashMap<>(Map.of());
+                // split the query string into key-value pairs
+                ObjectNode queryParamsNode = ParsingUtilities.mapper.createObjectNode();
                 if(qp.length() != 0) {
                     // get the query parameters as a list of name-value pairs
                     Arrays.stream(qp.split("&"))
-                            .forEach(pair -> queryParams.put(pair.split("=")[0], pair.split("=")[1]));
+                            .forEach(pair -> queryParamsNode.put(pair.split("=")[0], pair.split("=")[1]));
                 }
 
-                return ParsingUtilities.mapper.readTree(objectMapper.writeValueAsString(Map.of(
-                        "scheme", uri.getScheme() == null ? "" : uri.getScheme(),
-                        "host", uri.getHost() == null ? "" : uri.getHost(),
-                        "port", String.valueOf(uri.getPort() == -1 ? 80 : uri.getPort()),
-                        "path", uri.getPath() == null ? "" : uri.getPath(),
-                        "query", uri.getQuery() == null ? "" : uri.getQuery(),
-                        "authority", uri.getAuthority() == null ? "" : uri.getAuthority(),
-                        "fragment", fragment,
-                        "query_params", objectMapper.writeValueAsString(queryParams)
-                )));
+                ObjectNode uriNode = ParsingUtilities.mapper.createObjectNode();
+                uriNode.put("scheme", uri.getScheme());
+                uriNode.put("host", uri.getHost());
+                uriNode.put("port", String.valueOf(uri.getPort() == -1 ? 80 : uri.getPort()));
```

Comment:
I guess this fallback could introduce a wrong port: for instance, if parsing `https://foo.com/` then you will return 80 as port, which is incorrect since the implied port for HTTPS connections (443). For simplicity you could just forward `uri.getPort()` without trying to add more logic to it, I would say.

Reply:
I see 🤔 alright then

Reply:
> I guess this fallback could introduce a wrong port: for instance, if parsing `https://foo.com/` then you will return 80 as port, which is incorrect since the implied port for HTTPS connections (443). For simplicity you could just forward `uri.getPort()` without trying to add more logic to it, I would say.

@wetneb: good catch! I should've seen it… ;-) 

I wonder here if it wouldn’t be better to return 443 for https and 80 for http. Because if this information can be eluded at the full URL level, it always implies those values. But at the `.port` level, we have no context. If you want to know the port, "" says nothing, since it could be 80 or 443, like you toughtfully mentioned.

Anyway, I understand why "" can also be what the user would want to see… I guess that since we are working the data, and not interpreting it, "" might be closer to the data.

Regards, Antoine